File: src/implementation/data/ButtonData.js
class ButtonData {
    constructor(parentData) {
        this.activity = parentData?.activity;
        this.date = parentData?.date;
        this.start = parentData?.startDateTime;
        this.end = parentData?.endDateTime;
    }

    static fromActivity(parentData, activity) {
        const instance = new ButtonData(parentData);
        instance.activity = activity;
        instance.type = "SelectActivity";
        return instance;
    }
    static fromDate(parentData, date) {
        const instance = new ButtonData(parentData);
        instance.date = date.toISOString().split('T')[0]; // Format date as YYYY-MM-DD
        instance.type = "SelectDate";
        return instance;
    }
    static fromTime(parentData, startDateTime, endDateTime) {
        const instance = new ButtonData(parentData);
        instance.date = undefined
        instance.activity = undefined
        instance.start = startDateTime.toTimeString().slice(0, 5);
        instance.end = endDateTime.toTimeString().slice(0, 5);
        instance.type = "SelectTime";
        return instance;
    }

    toString() {
        return JSON.stringify(this);
    }
}

class ConfirmButtonData extends ButtonData {
    constructor(parentData) {
        super(parentData);
        this.type = "Confirm";
    }
}

File: src/Telegram/ButtonData.cs
using Newtonsoft.Json;
using TelegramBot.Core;

namespace TelegramBot.Telegram;

public record ButtonData(string type)
{
    public string Type { get; } = type;
    public int Activity { get; init; }
    public string? Date { get; init; }
    public string? Start { get; init; }
    public string? End { get; init; }
    public string? EventId { get; init; }

    public static ButtonData FromActivity(Activity activity)
    {
        return new ButtonData("SelectActivity") { Activity = (int)activity };
    }

    public static ButtonData FromDate(ButtonData parentData, DateOnly date)
    {
        return new ButtonData("SelectDate")
        {
            Activity = parentData.Activity, Date = date.ToString("yyyy-MM-dd")
        };
    }

    public static ButtonData FromTime(TimeOnly from, TimeOnly to)
    {
        return new ButtonData("SelectTime") { Start = from.ToString("HH:mm"), End = to.ToString("HH:mm") };
    }

    public string ToJson()
    {
        return JsonConvert.SerializeObject(this, JsonSetting.Minimal);
    }
}

public static class JsonSetting
{
    public static readonly JsonSerializerSettings Minimal = new()
    {
        NullValueHandling = NullValueHandling.Ignore,
        DefaultValueHandling = DefaultValueHandling.Ignore,
        Formatting = Formatting.None
    };

    public static readonly JsonSerializerSettings Intended = new()
    {
        NullValueHandling = NullValueHandling.Ignore,
        DefaultValueHandling = DefaultValueHandling.Ignore,
        Formatting = Formatting.Indented
    };
}
